### PR TITLE
Add redis cluster load balancing

### PR DESCRIPTION
**Pull request by HenryYYang**
Description: Use Redis Cluster load balancing for Redis clusters. 
Risk Level: low
Testing: Added unit tests and integration tests
Docs Changes: yes
Release Notes: yes

### CODE REVIEW

**Comment by alyssawilk** [comment]
Drive by comment - can we add an entry for redis cluster in CODEOWNERS as part of this work?
thanks!

**Comment by mattklein123** [comment]
@FAYiEKcbD0XFqF2QK2E4viAHg8rMm2VbjYKdjTg and @msukalski for first pass.

File: source/common/upstream/upstream_impl.h [hunk 1]
```diff
@@ -815,5 +819,10 @@ class BaseDynamicClusterImpl : public ClusterImplBase {
  */
 Network::DnsLookupFamily getDnsLookupFamilyFromCluster(const envoy::api::v2::Cluster& cluster);
 
+/**
+ * Utility function to get Dns from cluster.
+ */
+Network::DnsLookupFamily getDnsLookupFamilyFromCluster(const envoy::api::v2::Cluster& cluster);
```
Comment (msukalski) [inline source/common/upstream/upstream_impl.h]: declared twice on lines 820 and 825?

File: test/common/common/hash_test.cc [hunk 2]
```diff
@@ -27,6 +27,14 @@ TEST(Hash, murmurHash2_64) {
   EXPECT_EQ(6142509188972423790U, MurmurHash::murmurHash2_64(""));
 }
 
+TEST(Hash, crc16) {
+  EXPECT_EQ(44950, Crc16::crc16("foo"));
```
Comment (msukalski) [inline test/common/common/hash_test.cc]: for my info -- how did you calculate these CRC-16 values?

Comment (HenryYYang) [reply test/common/common/hash_test.cc]: I used the online CRC16 calculator: https://www.lammertbies.nl/comm/info/crc-calculation.html, then I cross check against a redis cluster using the CLUSTER KEYSLOT command to ensure we ended up with the same slot.

File: source/common/upstream/upstream_impl.h (hunk 1 above)
Comment (HenryYYang) [reply source/common/upstream/upstream_impl.h]: It's a merge error, thanks for pointing it out.

File: tools/spelling_dictionary.txt [hunk 3]
```diff
@@ -814,3 +814,5 @@ zag
 zig
 zlib
 zxid
+xmodern
```
[the remark this answers is not in the document]
Comment (HenryYYang) [reply tools/spelling_dictionary.txt]: thanks

File: test/common/common/hash_test.cc (hunk 2 above)
Comment (msukalski) [reply test/common/common/hash_test.cc]: excellent

**Comment by msukalski** [comment]
After looking at the check failures and the code, I assume that the failed checks are due to extra memory allocated for cluster_type_ in the ClusterInfoImpl. That's a little bit misleading as part of the stats integration testing, but I don't see how this extra overhead is not absorbed into the TCmalloc generic allocation.

**Comment by HenryYYang** [comment]
> After looking at the check failures and the code, I assume that the failed checks are due to extra memory allocated for cluster_type_ in the ClusterInfoImpl. That's a little bit misleading as part of the stats integration testing, but I don't see how this extra overhead is not absorbed into the TCmalloc generic allocation.
Yeah, I was having issues because I was running the tests in Clion and ipv4 and ipv6 were giving different memory.  I've figured it out and it's fixed now.  Thanks.

**Comment by msukalski** [comment]
LGTM.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.h [hunk 4]
```diff
@@ -66,6 +68,38 @@ class InstanceImpl : public Instance {
 
   typedef std::unique_ptr<ThreadLocalActiveClient> ThreadLocalActiveClientPtr;
 
+  class RequestRouter {
+  public:
+    virtual ~RequestRouter(){};
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.h]: nit: del trailing ';' same elsewhere

File: test/common/common/hash_test.cc (hunk 2 above)
Comment (mattklein123) [reply test/common/common/hash_test.cc]: Please add a comment to this effect.

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 5]
```diff
@@ -69,6 +114,18 @@ void InstanceImpl::ThreadLocalPool::onClusterAddOrUpdateNonVirtual(
   }
 
   ASSERT(cluster_ == nullptr);
+  if (cluster.info()->clusterType().name() == Extensions::Clusters::ClusterTypes::get().Redis) {
+    auto clusters = parent_.cm_.clusters();
+    auto clusterInstance = clusters.find(cluster_name_);
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: nit: `cluster_instance`

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 6]
```diff
@@ -69,6 +114,18 @@ void InstanceImpl::ThreadLocalPool::onClusterAddOrUpdateNonVirtual(
   }
 
   ASSERT(cluster_ == nullptr);
+  if (cluster.info()->clusterType().name() == Extensions::Clusters::ClusterTypes::get().Redis) {
+    auto clusters = parent_.cm_.clusters();
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: This call is not thread safe. `clusters()` should only be called on the main thread. Beyond thread safety, it's not safe since it's possible for the central cluster to get removed before the thread local cluster update goes out (if there is a quick add/remove sequence). We will need to figure out a better way to properly handle this. Can you take a look and see if you can come up with something and potentially come talk to me IRL if you want to brainstorm (I don't have any quick ideas I would need to sit and think about this)?

Comment (HenryYYang) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: Done

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 7]
```diff
@@ -37,6 +39,49 @@ InstanceImpl::makeRequestToHost(const std::string& host_address,
...
+  LbContextImpl lb_context(key, parent_.config_.enableHashtagging());
+  return cluster_.loadBalancer().chooseHost(&lb_context);
+}
+
+InstanceImpl::RedisClusterRouter::RedisClusterRouter(
+    InstanceImpl& parent, const Envoy::Extensions::Clusters::Redis::RedisCluster& redis_cluster)
+    : parent_(parent), redis_cluster_(redis_cluster) {}
+
+Upstream::HostConstSharedPtr InstanceImpl::RedisClusterRouter::chooseHost(const std::string& key) {
+  uint64_t slot = Crc16::crc16(hashtag(key, parent_.config_.enableHashtagging())) %
```
[the remark this answers is not in the document]
Comment (HenryYYang) [reply source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: done

File: docs/root/intro/version_history.rst [hunk 8]
```diff
@@ -66,6 +67,9 @@ Version history
   that allows ignoring new hosts for the purpose of load balancing calculations until they have
   been health checked for the first time.
 * upstream: added runtime error checking to prevent setting dns type to STRICT_DNS or LOGICAL_DNS when custom resolver name is specified.
+* upstream: added load balancer factory method on cluster, cluster implementation can now control the type of load balancer to be used.
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: I think there is a merge issue here. Delete.

File: source/common/upstream/upstream_impl.cc [hunk 9]
```diff
@@ -631,6 +631,30 @@ ClusterInfoImpl::ClusterInfoImpl(const envoy::api::v2::Cluster& config,
     eds_service_name_ = config.eds_cluster_config().service_name();
   }
 
+  if (!config.has_cluster_type()) {
+    switch (config.type()) {
+    case envoy::api::v2::Cluster::STATIC:
+      cluster_type_.set_name(Extensions::Clusters::ClusterTypes::get().Static);
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.cc]: Why is this needed? Perhaps just return an optional if you need access to custom cluster data and remove this switch statement?

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 10]
```diff
@@ -77,17 +79,21 @@ void RedisCluster::updateAllHosts(const Upstream::HostVector& hosts_added,
...
-    new_hosts.emplace_back(new RedisHost(info(), "", slot.master_, *this, true));
+    new_hosts.emplace_back(new RedisHost(info(), "", slot.master(), *this, true));
   }
 
   std::unordered_map<std::string, Upstream::HostSharedPtr> updated_hosts;
   Upstream::HostVector hosts_added;
   Upstream::HostVector hosts_removed;
-  if (updateDynamicHostList(new_hosts, hosts_, hosts_added, hosts_removed, updated_hosts,
-                            all_hosts_)) {
+  bool host_updated = updateDynamicHostList(new_hosts, hosts_, hosts_added, hosts_removed,
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.cc]: nit: const here and below.

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 11]
```diff
@@ -317,17 +314,27 @@ RedisClusterFactory::createClusterWithConfig(
     Envoy::Stats::ScopePtr&& stats_scope) {
   if (!cluster.has_cluster_type() ||
       cluster.cluster_type().name() != Extensions::Clusters::ClusterTypes::get().Redis) {
-    throw EnvoyException("Redis cluster can only created with redis cluster type");
+    throw EnvoyException("Redis cluster can only created with redis cluster type.");
+  }
+  // TODO(hyang): This is needed to migrate existing cluster, disallow using other lb_policy
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.cc]: This feature is not complete or fully working yet. I don't think we need to deal with back compat? I would just switch it?

File: source/extensions/clusters/redis/redis_cluster.cc [hunk 12]
```diff
@@ -77,17 +79,21 @@ void RedisCluster::updateAllHosts(const Upstream::HostVector& hosts_added,
...
   }
 
   std::unordered_map<std::string, Upstream::HostSharedPtr> updated_hosts;
   Upstream::HostVector hosts_added;
   Upstream::HostVector hosts_removed;
-  if (updateDynamicHostList(new_hosts, hosts_, hosts_added, hosts_removed, updated_hosts,
-                            all_hosts_)) {
+  bool host_updated = updateDynamicHostList(new_hosts, hosts_, hosts_added, hosts_removed,
+                                            updated_hosts, all_hosts_);
+  bool slot_updated = lb_factory_ ? lb_factory_->onClusterSlotUpdate(slots, updated_hosts) : false;
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.cc]: When is lb_factory_ null? You pass as part of the constructor?

File: source/extensions/clusters/redis/redis_cluster.h [hunk 13]
```diff
@@ -259,8 +248,7 @@ class RedisCluster : public Upstream::BaseDynamicClusterImpl {
   const LocalInfo::LocalInfo& local_info_;
   Runtime::RandomGenerator& random_;
   RedisDiscoverySession redis_discovery_session_;
-  // The slot to master node map.
-  SlotArray cluster_slots_map_;
+  ClusterSlotUpdateCallBackSharedPtr lb_factory_;
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster.h]: nit: const

File: source/extensions/clusters/redis/redis_cluster_lb.h [hunk 14]
```diff
@@ -0,0 +1,136 @@
...
+namespace Envoy {
+namespace Extensions {
+namespace Clusters {
+namespace Redis {
+
+static const uint64_t MaxSlot = 16384;
+
+typedef std::array<Upstream::HostSharedPtr, MaxSlot> SlotArray;
+
+typedef std::shared_ptr<SlotArray> SlotArraySharedPtr;
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster_lb.h]: make this `std::shared_ptr<const SlotArray> `

File: source/extensions/clusters/redis/redis_cluster_lb.h [hunk 15]
```diff
@@ -0,0 +1,136 @@
...
+ * `Redis Cluster Spec <https://redis.io/topics/cluster-spec#keys-distribution-model`_, the key
+ * space is split into a fixed size 16384 slots. The current implementation uses a fixed size
+ * std::array() of shared_ptr pointing to the master host. This has a fixed cpu and memory cost and
+ * provide a fast lookup constant time lookup similar to Maglev. This will be used by the redis
+ * proxy filter for load balancing purpose.
+ */
+class RedisClusterLoadBalancer : public Upstream::LoadBalancer {
+public:
+  RedisClusterLoadBalancer(SlotArraySharedPtr slot_array) : slot_array_(std::move(slot_array)){};
+  ~RedisClusterLoadBalancer() = default;
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster_lb.h]: delete

File: source/extensions/clusters/redis/redis_cluster_lb.h [hunk 16]
```diff
@@ -0,0 +1,136 @@
...
+typedef std::shared_ptr<ClusterSlotUpdateCallBack> ClusterSlotUpdateCallBackSharedPtr;
+
+/**
+ * This factory is created and returned by RedisCluster's factory() method, the create() method will
+ * be called on each thread to create a thread local RedisClusterLoadBalancer.
+ */
+class RedisClusterLoadBalancerFactory : public ClusterSlotUpdateCallBack,
+                                        public Upstream::LoadBalancerFactory {
+public:
+  RedisClusterLoadBalancerFactory() = default;
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster_lb.h]: delete

File: source/extensions/clusters/redis/redis_cluster_lb.cc [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+namespace Extensions {
+namespace Clusters {
+namespace Redis {
+
+// RedisClusterLoadBalancerFactory
+bool RedisClusterLoadBalancerFactory::onClusterSlotUpdate(
+    const std::vector<Envoy::Extensions::Clusters::Redis::ClusterSlot>& slots,
+    Envoy::Upstream::HostMap all_hosts) {
+
+  absl::WriterMutexLock lock(&mutex_);
```
Comment (mattklein123) [inline source/extensions/clusters/redis/redis_cluster_lb.cc]: You only need to hold the writer lock write before you update the member shared pointer value.

File: source/extensions/clusters/redis/redis_cluster.cc (hunk 11 above)
Comment (HenryYYang) [reply source/extensions/clusters/redis/redis_cluster.cc]: We're currently using it with Random, so if we don't do back compat then we'll have to make sure both the config changes and the envoy binary are roll out at the same time.  It's better to do this and remove it after both the config changes and binary are roll out.

File: source/extensions/clusters/redis/redis_cluster.cc (hunk 12 above)
Comment (HenryYYang) [reply source/extensions/clusters/redis/redis_cluster.cc]: This is due to the back compat issue, this can be removed after the back compat is used.

File: docs/root/intro/version_history.rst (hunk 8 above)
Comment (HenryYYang) [reply docs/root/intro/version_history.rst]: done

File: source/extensions/clusters/redis/redis_cluster_lb.cc (hunk 17 above)
Comment (HenryYYang) [reply source/extensions/clusters/redis/redis_cluster_lb.cc]: that's what I had before, but clang treat it as an error.  Is there a way to disable that check here? Alternatively I do a reader lock first, then a writer lock when we actually do the update?

File: docs/root/intro/version_history.rst [hunk 18]
```diff
@@ -66,6 +67,8 @@ Version history
   that allows ignoring new hosts for the purpose of load balancing calculations until they have
   been health checked for the first time.
 * upstream: added runtime error checking to prevent setting dns type to STRICT_DNS or LOGICAL_DNS when custom resolver name is specified.
+* grpc-json: added support for :ref:`auto mapping
```
Comment (mattklein123) [inline docs/root/intro/version_history.rst]: merge issue

File: include/envoy/upstream/upstream.h [hunk 19]
```diff
@@ -724,6 +724,13 @@ class ClusterInfo {
    */
   virtual envoy::api::v2::Cluster::DiscoveryType type() const PURE;
 
+  /**
+   * @return the type of cluster, this includes custom discovery type in addition to the build-in
```
Comment (mattklein123) [inline include/envoy/upstream/upstream.h]: Is this comment still accurate?

File: source/common/upstream/upstream_impl.cc [hunk 20]
```diff
@@ -574,7 +574,8 @@ ClusterInfoImpl::ClusterInfoImpl(const envoy::api::v2::Cluster& config,
       cluster_socket_options_(parseClusterSocketOptions(config, bind_config)),
       drain_connections_on_host_removal_(config.drain_connections_on_host_removal()),
       warm_hosts_(!config.health_checks().empty() &&
-                  common_lb_config_.ignore_new_hosts_until_first_hc()) {
+                  common_lb_config_.ignore_new_hosts_until_first_hc()),
+      cluster_type_(config.cluster_type()) {
```
Comment (mattklein123) [inline source/common/upstream/upstream_impl.cc]: This optional is always going to have a value. Is that what you want? I don't think so?

File: source/extensions/clusters/redis/crc16.cc [hunk 21]
```diff
@@ -0,0 +1,55 @@
...
+    0x34e2, 0x24c3, 0x14a0, 0x0481, 0x7466, 0x6447, 0x5424, 0x4405, 0xa7db, 0xb7fa, 0x8799, 0x97b8,
+    0xe75f, 0xf77e, 0xc71d, 0xd73c, 0x26d3, 0x36f2, 0x0691, 0x16b0, 0x6657, 0x7676, 0x4615, 0x5634,
+    0xd94c, 0xc96d, 0xf90e, 0xe92f, 0x99c8, 0x89e9, 0xb98a, 0xa9ab, 0x5844, 0x4865, 0x7806, 0x6827,
+    0x18c0, 0x08e1, 0x3882, 0x28a3, 0xcb7d, 0xdb5c, 0xeb3f, 0xfb1e, 0x8bf9, 0x9bd8, 0xabbb, 0xbb9a,
+    0x4a75, 0x5a54, 0x6a37, 0x7a16, 0x0af1, 0x1ad0, 0x2ab3, 0x3a92, 0xfd2e, 0xed0f, 0xdd6c, 0xcd4d,
+    0xbdaa, 0xad8b, 0x9de8, 0x8dc9, 0x7c26, 0x6c07, 0x5c64, 0x4c45, 0x3ca2, 0x2c83, 0x1ce0, 0x0cc1,
+    0xef1f, 0xff3e, 0xcf5d, 0xdf7c, 0xaf9b, 0xbfba, 0x8fd9, 0x9ff8, 0x6e17, 0x7e36, 0x4e55, 0x5e74,
+    0x2e93, 0x3eb2, 0x0ed1, 0x1ef0};
+
+// (https://github.com/antirez/redis/blob/unstable/src/crc16.c)
```
Comment (mattklein123) [inline source/extensions/clusters/redis/crc16.cc]: Can you move this to the top of the value and be very clear in a comment about where all of this code is copied from?

File: source/extensions/filters/network/redis_proxy/conn_pool_impl.cc [hunk 22]
```diff
@@ -155,7 +155,13 @@ InstanceImpl::ThreadLocalPool::makeRequest(const std::string& key,
     return nullptr;
   }
 
-  LbContextImpl lb_context(key, parent_.config_.enableHashtagging());
+  Upstream::ClusterInfoConstSharedPtr info = cluster_->info();
+  const auto& clusterType = info->clusterType();
```
Comment (mattklein123) [inline source/extensions/filters/network/redis_proxy/conn_pool_impl.cc]: nit: `cluster_type`
